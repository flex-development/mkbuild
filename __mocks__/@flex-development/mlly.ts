/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
