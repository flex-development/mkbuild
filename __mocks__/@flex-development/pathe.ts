/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
