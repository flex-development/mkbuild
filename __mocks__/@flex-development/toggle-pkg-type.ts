/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
