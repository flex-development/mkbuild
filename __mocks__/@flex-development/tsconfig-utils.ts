/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
