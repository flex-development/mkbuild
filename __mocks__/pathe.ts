/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
