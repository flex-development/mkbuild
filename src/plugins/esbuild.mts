/**
 * @file Plugins - esbuild
 * @module mkbuild/plugins/esbuild
 */

import formatEsbuildMessage from '#internal/format-esbuild-message'
import type WithUndefined from '#types/with-undefined'
import type {
  EsbuildOptions,
  Format
} from '@flex-development/mkbuild'
import pathe from '@flex-development/pathe'
import { ifelse, isNIL } from '@flex-development/tutils'
import { createFilter } from '@rollup/pluginutils'
import { ok } from 'devlop'
import type * as Esbuild from 'esbuild'
import type { Loader, LogLevel, TransformOptions, TsconfigRaw } from 'esbuild'
import type {
  MinimalPluginContext,
  ModuleInfo,
  NormalizedInputOptions,
  NormalizedOutputOptions,
  OutputBundle,
  Plugin,
  PluginContext,
  ProgramNode,
  RenderedChunk,
  RollupLog,
  LogLevel as RollupLogLevel,
  SourceDescription,
  TransformPluginContext,
  TransformResult
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:esbuild'

/**
 * esbuild module.
 *
 * @internal
 *
 * @var {typeof Esbuild | undefined} esbuild
 */
let esbuild: typeof Esbuild | undefined

plugin.PLUGIN_NAME = PLUGIN_NAME
export default plugin

/**
 * Plugin options.
 *
 * @internal
 *
 * @extends {EsbuildOptions}
 */
interface Options extends EsbuildOptions {
  /**
   * Module format.
   *
   * @see {@linkcode Format}
   */
  format: Format

  /**
   * Set the value of the `sourceRoot` field in a source map, which specifies
   * the path that all other paths in the source map are relative to. If this
   * field is not present, all paths in the source map are interpreted as being
   * relative to the directory containing the source map instead.
   *
   * @see https://esbuild.github.io/api/#source-root
   */
  sourceRoot?: URL | string | null | undefined

  /**
   * Include the `sourcesContent` field in sourcemaps.
   *
   * @see https://esbuild.github.io/api/#sources-content
   */
  sourcesContent?: boolean | null | undefined
}

/**
 * Create a plugin to transform files with esbuild.
 *
 * @see {@linkcode Options}
 * @see {@linkcode Plugin}
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Options} [opts]
 *  esbuild options
 * @return {Plugin}
 *  esbuild transform plugin
 */
function plugin(this: void, opts: Options): Plugin {
  /**
   * Copied modules.
   *
   * @const {Set<ModuleInfo>} copied
   */
  const copied: Set<ModuleInfo> = new Set<ModuleInfo>()

  /**
   * Module id filter.
   *
   * @const {(id: unknown) => boolean} filter
   */
  const filter: (id: unknown) => boolean = createFilter(
    opts.include ?? /\.(?:(?:json)|(?:[cm]?[jt]s)|(?:[jt]sx?))$/,
    opts.exclude ?? undefined
  )

  /**
   * Raw tsconfig.
   *
   * @var {TsconfigRaw} tsconfigRaw
   */
  let tsconfigRaw: TsconfigRaw

  return {
    buildStart,
    generateBundle,
    name: PLUGIN_NAME,
    onLog,
    renderChunk,
    transform
  }

  /**
   * Get the raw tsconfig from `options`.
   *
   * @this {PluginContext}
   *
   * @param {NormalizedInputOptions} options
   *  Normalized input options
   * @return {undefined}
   */
  function buildStart(
    this: PluginContext,
    options: NormalizedInputOptions
  ): undefined {
    ok(typeof options.tsconfig === 'object', 'expected `options.tsconfig`')
    tsconfigRaw = options.tsconfig as TsconfigRaw
    return void this
  }

  /**
   * Add the content of copied modules to their chunks. Any sourcemaps generated
   * for copied modules will be removed.
   *
   * @see https://rollupjs.org/plugin-development/#generatebundle
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} options
   *  Normalized output options
   * @param {OutputBundle} bundle
   *  Output bundle
   * @return {undefined}
   */
  function generateBundle(
    this: PluginContext,
    options: NormalizedOutputOptions,
    bundle: OutputBundle
  ): undefined {
    for (const output of Object.values(bundle)) {
      for (const info of copied) {
        if (output.type === 'chunk' && output.facadeModuleId === info.id) {
          ok(typeof info.code === 'string', 'expected `info.code`')

          output.code = info.code
          output.map = null

          if (output.sourcemapFileName !== null) {
            delete bundle[output.sourcemapFileName]
            output.sourcemapFileName = null
          }
        }
      }
    }

    return void bundle
  }

  /**
   * Skip `EMPTY_BUNDLE` logs.
   *
   * @see https://rollupjs.org/plugin-development/#onlog
   *
   * @this {MinimalPluginContext}
   *
   * @param {RollupLogLevel} level
   *  Rollup log level
   * @param {RollupLog} log
   *  Rollup log object
   * @return {false | undefined}
   *  `false` if log will be silenced, `undefined` otherwise
   */
  function onLog(
    this: MinimalPluginContext,
    level: RollupLogLevel,
    log: RollupLog
  ): false | undefined {
    return ifelse(log.code === 'EMPTY_BUNDLE', false, void log)
  }

  /**
   * Fix the name and file name of copied modules.
   *
   * @see https://rollupjs.org/plugin-development/#renderchunk
   *
   * @this {PluginContext}
   *
   * @param {string} code
   *  Chunk content
   * @param {RenderedChunk} chunk
   *  The chunk to process
   * @return {undefined}
   */
  function renderChunk(
    this: PluginContext,
    code: string,
    chunk: RenderedChunk
  ): undefined {
    for (const info of copied) {
      if (chunk.facadeModuleId === info.id) {
        chunk.fileName = pathe.changeExt(chunk.fileName, pathe.extname(info.id))
        chunk.name = pathe.changeExt(chunk.fileName, null)
        break
      }
    }

    return void chunk
  }

  /**
   * Transform `code` using the [esbuild transform api][esbuild-transform].
   *
   * [esbuild-transform]: https://esbuild.github.io/api/#transform
   *
   * @see https://rollupjs.org/plugin-development/#transform
   *
   * @todo `globalName`
   * @todo `ignoreAnnotations`
   * @todo `pure`
   * @todo `treeShaking`
   *
   * @this {TransformPluginContext}
   *
   * @param {string} code
   *  The code to transform
   * @param {string} id
   *  The id of the module containing `code`
   * @return {Promise<TransformResult>}
   *  Transform result
   */
  async function transform(
    this: TransformPluginContext,
    code: string,
    id: string
  ): Promise<TransformResult> {
    /**
     * Transform result.
     *
     * @var {WithUndefined<Partial<SourceDescription>> | undefined} result
     */
    let result: WithUndefined<Partial<SourceDescription>> | undefined

    if (filter(id)) {
      /**
       * AST of transformed code.
       *
       * @var {ProgramNode | undefined} ast
       */
      let ast: ProgramNode | undefined

      /**
       * Content loader.
       *
       * @var {Loader | undefined} loader
       */
      let loader: Loader | undefined = pathe.extToValue(pathe.basename(id), {
        '.cjs': 'js',
        '.cts': 'ts',
        '.d.cts': 'copy',
        '.d.mts': 'copy',
        '.d.ts': 'copy',
        '.js': 'js',
        '.json': 'copy',
        '.jsx': 'jsx',
        '.mjs': 'js',
        '.mts': 'ts',
        '.ts': 'ts',
        '.tsx': 'tsx',
        ...opts.loader
      })

      // note: rather than just be a no-op, or work as `empty`, esbuild throws
      // an annoying error when the `copy` loader is used with `transform`.
      if (loader === 'copy') {
        /**
         * Module info.
         *
         * @const {ModuleInfo | null} info
         */
        const info: ModuleInfo | null = this.getModuleInfo(id)

        if (info) {
          copied.add(info)
          info.code = code
        }

        ast = {
          body: [],
          end: 0,
          sourceType: 'module',
          start: 0,
          type: 'Program'
        }

        loader = 'empty'
      }

      const { transform } = esbuild ??= await import('esbuild')

      /**
       * Log level overrides.
       *
       * @const {Record<string, LogLevel>} logOverride
       */
      const logOverride: Record<string, LogLevel> = { ...opts.logOverride }

      ok(typeof opts.sourcesContent === 'boolean', 'expected `sourcesContent`')

      const {
        code: transformed,
        map,
        legalComments,
        mangleCache,
        warnings
      } = await transform(code, {
        charset: opts.charset ?? undefined,
        color: false,
        define: opts.define ?? undefined,
        drop: opts.drop ?? undefined,
        dropLabels: opts.dropLabels ?? undefined,
        format: opts.format,
        jsx: opts.jsx ?? undefined,
        jsxDev: opts.jsxDev ?? undefined,
        jsxFactory: opts.jsxFactory ?? undefined,
        jsxFragment: opts.jsxFragment ?? undefined,
        jsxImportSource: opts.jsxImportSource ?? undefined,
        jsxSideEffects: opts.jsxSideEffects ?? undefined,
        keepNames: opts.keepNames ?? undefined,
        legalComments: 'external',
        lineLimit: opts.lineLimit ?? undefined,
        loader,
        logLevel: 'silent',
        logLimit: 0,
        logOverride: Object
          .entries(logOverride)
          .reduce<Record<string, LogLevel>>((acc, [key, level]) => {
            acc[key] = ifelse(level === 'silent', level, 'warning')
            return acc
          }, {}),
        mangleCache: opts.mangleCache ?? undefined,
        mangleProps: opts.mangleProps ?? undefined,
        mangleQuoted: opts.mangleQuoted ?? undefined,
        minify: opts.minify ?? undefined,
        minifyIdentifiers: opts.minifyIdentifiers ?? undefined,
        minifySyntax: opts.minifySyntax ?? undefined,
        minifyWhitespace: opts.minifyWhitespace ?? undefined,
        platform: opts.platform ?? undefined,
        reserveProps: opts.reserveProps ?? undefined,
        sourceRoot: ifelse(
          isNIL(opts.sourceRoot),
          undefined,
          String(opts.sourceRoot)
        ),
        sourcefile: id,
        sourcemap: true,
        sourcesContent: opts.sourcesContent,
        supported: {
          'dynamic-import': true,
          'import-meta': true,
          ...opts.supported
        },
        target: typeof opts.target === 'object'
          ? [...new Set(opts.target)]
          : opts.target,
        tsconfigRaw: opts.tsconfigRaw ?? tsconfigRaw
      } as TransformOptions)

      for (const message of warnings) {
        /**
         * Message formatted as rollup log.
         *
         * @const {RollupLog} log
         */
        const log: RollupLog = formatEsbuildMessage(
          message,
          code,
          id,
          PLUGIN_NAME,
          logOverride[message.id]
        )

        // log message if not silenced
        log.level && this[log.level](log, log.loc)
      }

      result = {
        ast,
        code: typeof ast === 'object' ? code : transformed,
        map,
        meta: { legalComments, mangleCache, warnings }
      }
    }

    return result as TransformResult
  }
}
