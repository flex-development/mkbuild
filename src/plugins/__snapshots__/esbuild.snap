// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:plugins/esbuild > should return plugin 1`] = `
{
  "buildStart": [Function buildStart],
  "generateBundle": [Function generateBundle],
  "name": "mkbuild:esbuild",
  "onLog": [Function onLog],
  "renderChunk": [Function renderChunk],
  "transform": [Function transform],
}
`;
